tokenprovider/token_provider.go

```diff
@@ -0,0 +1,66 @@
+package tokenprovider
+
+import (
+	"fmt"
+	"sort"
+	"strings"
+	"sync"
+	"time"
+
+	"golang.org/x/oauth2"
+)
+
+var (
+	tokenCache = oauthTokenCacheType{
+		cache: map[string]*oauth2.Token{},
+	}
+
+	// timeNow makes it possible to test usage of time
+	timeNow = time.Now
+)
+
+type oauthTokenCacheType struct {
+	cache map[string]*oauth2.Token
+	sync.Mutex
+}
+
+// TokenProvider is anything that can return a token
+type TokenProvider interface {
+	GetAccessToken() (string, error)
+}
+
+// tokenProviderImpl implements the TokenProvider interface
+type tokenProviderImpl struct {
+	cacheKey    string
+	tokenSource oauth2.TokenSource
+}
+
+// GetAccessToken implements TokenProvider
+func (tpi *tokenProviderImpl) GetAccessToken() (string, error) {
```

Comment:
I'm not a golang person, but I find this pattern of short and obtuse variable names in grafana's go code to be really hard to read.  

Perhaps this could be like `tokenProvider` or something else that makes it more readable at first glance?